### PR TITLE
needless_collect: catch x: Vec<_> = iter.collect(); x.into_iter() ...

### PR DESCRIPTION
**Pull request by JarredAllen**
changelog: Expand the needless_collect lint as suggested in #5627 (WIP).

This PR is WIP because I can't figure out how to make the multi-part suggestion include its changes in the source code (the fixed is identical to the source, despite the lint making suggestions). Aside from that one issue, I think this should be good.

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -2425,6 +2429,158 @@ fn check_needless_collect<'tcx>(expr: &'tcx Expr<'_>, cx: &LateContext<'tcx>) {
...
+            if let Some(Expr { kind: ExprKind::Path(QPath::Resolved(_, ref path)), .. }) = args.get(0);
+            if let &[name] = &path.segments;
+            if name.ident == self.target;
+            then {
+                let into_iter = sym!(into_iter);
+                let len = sym!(len);
+                let is_empty = sym!(is_empty);
+                let contains = sym!(contains);
+                match method_name.ident.name {
+                    name if name == into_iter => self.uses.push(
```
Comment (phansch) [inline clippy_lints/src/loops.rs:2532]: Ah.. function calls are not allowed in patterns. At least for `into_iter` we can use the pre-interned symbol from rustc (will need a `use rustc_span::symbol::sym`):

https://github.com/rust-lang/rust/blob/46cf80dc1a75ad27f67e79f73fec371a16762494/src/librustc_span/symbol.rs#L594

```suggestion
                    sym::into_iter => self.uses.push(
```

File: clippy_lints/src/loops.rs [hunk 2]
```diff
@@ -2425,6 +2429,158 @@ fn check_needless_collect<'tcx>(expr: &'tcx Expr<'_>, cx: &LateContext<'tcx>) {
...
+                if let StmtKind::Local(
+                    Local { pat: Pat { kind: PatKind::Binding(_, _, ident, .. ), .. },
+                    init: Some(ref init_expr), .. }
+                ) = stmt.kind;
+                if let ExprKind::MethodCall(ref method_name, _, &[ref iter_source], ..) = init_expr.kind;
+                if method_name.ident.name == sym!(collect) && match_trait_method(cx, &init_expr, &paths::ITERATOR);
+                if let Some(ref generic_args) = method_name.args;
+                if let Some(GenericArg::Type(ref ty)) = generic_args.args.get(0);
+                if let ty = cx.typeck_results().node_type(ty.hir_id);
+                if is_type_diagnostic_item(cx, ty, sym!(vec_type)) ||
```
Comment (phansch) [inline clippy_lints/src/loops.rs:2445]:
```suggestion
                if is_type_diagnostic_item(cx, ty, sym::vec_type) ||
```

**Comment by JarredAllen** [comment]
@phansch I think I've added those suggestions.

**Pull request merged**